src/pulsar_producer.erl

```diff
@@ -355,19 +348,39 @@ do_connect(State = #{opts := Opts, broker_server := {Host, Port}}) ->
              [{state_timeout, ?RECONNECT_TIMEOUT, do_connect}]}
     end.
 
-code_change({down, _Vsn} = Direction, State, Data0, _Extra) ->
+code_change({down, _ToVsn}, State, Data, Extra) when is_map(Data) ->
+    #{to_version := ToVsn} = Extra,
+    case pulsar_relup:is_before_replayq(ToVsn) of
+        true ->
+            do_code_change_down_replayq(State, Data, Extra);
+        false ->
+            {ok, State, Data}
+    end;
+code_change({down, _Vsn}, State, Data, _Extra) ->
+    {ok, State, Data};
+code_change(_ToVsn, State, DataRec, Extra) when is_tuple(DataRec) ->
+    #{to_version := ToVsn} = Extra,
+    case pulsar_relup:is_before_replayq(ToVsn) of
+        true ->
+            {ok, State, DataRec};
+        false ->
+            do_code_change_up_replayq(State, DataRec, Extra)
```

Comment:
we'll need to check from which version it's upgrading?
in case `code_change` going to be used for a version after `0.7.0`
however, that's a future problem, we don't have to add such code now.

Reply:
In this clause, we have a [`when is_tuple(DataRec)` ](https://github.com/emqx/pulsar-client-erl/pull/42/files#diff-2e29e17e5c895f84ea4e060ec5bbf3ecd713941ebe435d097518d263e6a9e19aR361-R363) that should match only when the state is in the old record format.  If in the future we upgrade, say, from 0.7.0 to 0.7.1, the state will already be a map, so it shouldn't fall into this clause.